diff --git a/kernel/pkg_observer.c b/kernel/pkg_observer.c
index 0fba687f..4547ae4f 100644
--- a/kernel/pkg_observer.c
+++ b/kernel/pkg_observer.c
@@ -21,24 +21,26 @@ struct watch_dir {
 
 static struct fsnotify_group *g;
 
-static int ksu_handle_inode_event(struct fsnotify_mark *mark, u32 mask,
-				  struct inode *inode, struct inode *dir,
-				  const struct qstr *file_name, u32 cookie)
+static int ksu_handle_inode_event(struct fsnotify_group *group,
+                                  struct inode *inode,
+                                  u32 mask, const void *data, int data_type,
+                                  const unsigned char *file_name, u32 cookie,
+                                  struct fsnotify_iter_info *iter_info)
 {
 	if (!file_name)
 		return 0;
 	if (mask & FS_ISDIR)
 		return 0;
-	if (file_name->len == 13 &&
-	    !memcmp(file_name->name, "packages.list", 13)) {
-		pr_info("packages.list detected: %d\n", mask);
-		track_throne();
-	}
+        size_t len = strlen((const char *)file_name);
+        if (len == 13 && memcmp(file_name, "packages.list", 13) == 0) {
+            pr_info("packages.list detected: %d\n", mask);
+            track_throne();
+        }
 	return 0;
 }
 
 static const struct fsnotify_ops ksu_ops = {
-	.handle_inode_event = ksu_handle_inode_event,
+	.handle_event = ksu_handle_inode_event,
 };
 
 static int add_mark_on_inode(struct inode *inode, u32 mask,
